ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1881,18 +2057,318 @@ void shutdown() throws InterruptedException {
     } finally {
       lock.unlock();
     }
+    shouldTerminateCurrentCycle = true;
     if (networkClient != null) {
       networkClient.close();
     }
     shutdownLatch.await();
   }
 
+  class RemoteReplicaGroupPoller {
```

Comment:
I would love to add more comment to explain the life cycle of a group.

I suppose it works like this
1. We preassign group id to each group. 
2. Once a group finishes one iteration of work, include metadata exchange and data download, we remove group id from the inflight group map and potentially backoff. If this group need to go through another iteration, the same group would be added back to inflight group map.
3. PollGroups would try to create (and add back) those groups.
4. When there is a new replica added to the thread, or a replica removed from this thread, we don't create new groups, just return the existing ones, and after they all finish current iteration, inflightGroupMap would become empty and do-while loop would break.
5. Or there is one group reaches max iteration.

Reply:
Instead of creating new groups, is it possible to reuse a group by just resetting its state to the start state ?

Reply:
Replicas can change in the group in each iteration, as some replicas may go to backoff mode 
or vice-versa.

Reply:
added comments.